### DISCUSSION TITLE
Feature request: The Search Dialog as a side panel

### DISCUSSION THREAD

**Discussion by wordmuse** [freeplane/freeplane#360] [Ideas]
I like the Search Dialog that lets me see all the instances of a word or phrase in a large mindmap. But it takes a good chunk of time to load in a large mindmap, and when it appears, it blocks a good chunk of the mindmap. Clicking **Go To** takes me to a given instance of the search results without dismissing the dialog, but then it continues to block the mindmap. Double-clicking a search result takes me to the chosen node and dismisses the dialog. The problem is that if I want to call the dialog again, I have to wait a bit.

My suggestion is that it be a possible side panel, in a similar place where a node's Note or its Properties panel are displayed. This would let me click an instance, keep the dialog open, and keep the dialog out of the way while it's open.

Thoughts?

~ Bal

**Comment by dpolivaev** [comment]
If Freeplane kept the dialog or the side panel it had to be synchronized with all changes in the map including added and deleted nodes and changed formula values. In other words, it would become a new way to see and probably edit the mind map.

I do not say it is not possible, it is just not easy. I need to think about it.

Who else needs this feature?

**Comment by macmarrum** [comment]
When I use the `Find and replace aka Search dialog`, it is synchronized for existing nodes, i.e. changes to text, details or note are reflected in the dialog (an icon change is not reflected). When I add, delete or move a node, the dialog is closed automatically.

I can make the dialog as wide as my side panel and position it so that it covers the side panel. When I close and reopen `Find and replace`, it pops up at the same position on the screen, giving me the feel of a side panel.

When I use `Ctrl+click` on a search result, the node is selected but the dialog stays open.

I rarely use `Find and replace`. I'd use it more often if I could limit its scope to the initial selection, so that I could search / replace a subset of the nodes. This makes me realize that I don't need `Find and replace` to become a (synchronized) side panel.

**Reply by euu2021** [reply]
> I rarely use Find and replace. I'd use it more often if I could limit its scope to the initial selection, so that I could search / replace a subset of the nodes.

If you apply a filter first, then you limit the scope.

**Reply by macmarrum** [reply]
Thanks for the tip, @eu2021. Very useful 👍

**Comment by wordmuse** [comment]
I just tried macmarrum's idea of changing the width of the dialog. It's close enough to what I want. Since it's not easy to do what I suggest, I'm going to withdraw my suggestion. I'm good.  Thanks guys...

~ Bal

**Comment by wordmuse** [comment]
One more thought. This time about the dialog. If there was a way to force it to stay open until I intentionally closed it, that would be super. Call this a nice-to-have rather than a full blown feature request. Only if it's easy to do - sort of like pinning it to be open. Otherwise, as I said, I'm good. 

~ Bal

**Reply by macmarrum** [reply]
I know of only 2 reasons the dialog is closed, other than closing it intentionally:
1. When I double-click a search result (goto and close)
2. When I change the map tree structure
  a. add, delete or move a node
  b. switch to a different tab

(1) can be avoided by holding `Ctrl`
(2) is probably the result of the dialog realizing it is out of date

@wordmuse which of the points are the most inconvenient for you: 1, 2a, 2b?

**Reply by dpolivaev** [reply]
@wordmuse Could you please answer this excellent question? It would help to understand your needs better.

**Reply by wordmuse** [reply]
I thought I did below. But sure...

1 - Double-clicking - this isn't much of a problem since, as was mentioned, it can be avoided by holding down the Ctrl key. I would prefer to reverse this, however: let double-clicking keep the dialog open and have Ctrl+double-clicking close it.

2a and 2b - both situations.  

But there is a third situation, not mentioned.

3 - when I call the dialog without having made any structural change or without changing tabs. It takes 23 seconds in my giant mindmap for the dialog to appear. In smaller mindmaps, the dialog comes up almost instantaneously. So it seems to be the size of the mindmap that is key here.

Is this helpful?

~ Bal

**Reply by dpolivaev** [reply]
How many nodes does your giant mind map have? (File->Map statistics…)
Does it use any formulas? Details? Notes? Formatted text in the core?
Could you send it to me so that I can try to optimize Freeplane so that it works with your map faster?

**Reply by wordmuse** [reply]
Dimitry - Here is a screenshot of my mindmap's statistics.

I don't use any formulas. There might be a few notes, but I prefer the details feature most of the time. I do format my text, primarily with bold and italics. I use underlines and strikethroughs sparingly.

I'm afraid that I'm not up for sending the file itself. It's a very personal journal that I'm not up showing anyone else, other than my wife. But I might have a way that I can give you something similar. Let me play around with things and see if what I'm thinking of can work.

~ Bal

![bwm-statistics](https://user-images.githubusercontent.com/8197883/168468056-d5c77c71-4384-43d2-b08e-4025d3105b83.png)


.

**Reply by wordmuse** [reply]
OK - My idea will pan out. I just have to do it. I can send you a sterilized version of my file. I can change all the letters into the letter _x_ using the Find and Replace dialog. It stays open for the procedure. The file size should not change as a result of these simple substitutions.

One thing I forgot to mention: I make extensive use of custom styles and custom icons. The styles will transfer, but not the icons, since they would need to reside on your local drive in the same path as what is on my computer.  I should be able to give you this file sometime later on today.

~ Bal

**Reply by dpolivaev** [reply]
Yes, please do it, it could help

**Reply by wordmuse** [reply]
Should I just attach it to a comment here, or would you prefer me to send it directly to you in some other way?

**Reply by dpolivaev** [reply]
Here is probably the easiest

**Reply by wordmuse** [reply]
OK - here you go. Github wouldn't let me attach a file with a _.mm_ extension, so I changed the extension to _.txt_. Just change it back to _.mm_ after you download it. Feel free to ask any questions you have about it.

~ Bal

[ipsolorem.txt](https://github.com/freeplane/freeplane/files/8695955/ipsolorem.txt)
.

**Comment by wordmuse** [comment]
I'm not sure your (2) is accurate. If I call the dialog, close it and call it again, I have the exact same wait time, each time. (I timed it: 23 seconds in my giant mindmap) And if I make a change, it's still the same 23 seconds because the change in the map is so small. Even if I make extensive changes, the changes are small relative to the size of the mindmap. So it seems to be dependent on the size of the map. FP seems to need to brute force some sort of indexing each time the dialog is called.

It's this time delay in my very large map that brought me here to make my request.

~ Bal

**Comment by dpolivaev** [comment]
I removed some cases where the dialog became closed in preview 1.10.3_01 except for cases when nodes get deleted or inserted. Such cases are more complicated but maybe I implement them later.

Consider also the new check box "Close after selection" at the bottom of the dialog. It replaces the not obvious holding of the CONTROL / META key.

**Reply by wordmuse** [reply]
Cool - I'll check it out later on today. Thanks much. :) 

~ Bal

**Reply by wordmuse** [reply]
I like the changes - thank you!!

**Comment by dpolivaev** [comment]
I used the opportunity to check why loading the search dialog takes so long on your map.
And it turned out that the time is mostly consumed when Freeplane identifies which icons should be shown because some icons come with styles and conditional styles should be calculated for all nodes. I can see different ways to handle it:

- I can introduce an option not to show the icons
- I can show only icons added to the node ignoring all icons coming with the style
- I can introduce an option not to show the icons coming with the style

Which solution would make more sense?

@wordmuse

**Reply by macmarrum** [reply]
I vote for "introduce an option not to show the icons".
This would give me the option to have less visual distractions in the dialog: no icons, just text.

**Comment by wordmuse** [comment]
I can't think of a time where I paid any attention to the icons in that dialog. In fact, when I resize the dialog, as described above, the icons are not in view. If the icons are slowing things down, then by all means, let's lose them. Maybe let there be an option (a box you can check) that says "show icons." I'm guessing very few people would click it - especially for large maps if they knew the effect it would have on speed.

Thanks, Dimitry!

~ Bal
